### DISCUSSION TITLE
Add ability to dehyrate & hydrate state (for integrating with TanStack Router)

### DISCUSSION THREAD

**Discussion by phuctm97** [pmndrs/jotai#2893] [Ideas]
TanStack Router/Start has built-in dehydrate & hydrate workflow for making state manager works very well with SSR. It would be great if Jotai supports it out-of-the-box as well.

```tsx
export function createRouter() {
  const store = createStore();
  const queryClient = new QueryClient();

  return createRouter({
    routeTree,
    context: {
      store,
      queryClient,
    },
    dehydrate: () => ({
      storeState: store.dehydrate(),
      queryClientState: dehydrate(queryClient),
    }),
    hydrate: ({ storeState, queryClientState }) => {
      store.hydrate(storeState);
      hydrate(queryClient, queryClientState)
    },
    Wrap: ({ children }) => (
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    ),
  })
}
```

**Comment by dai-shi** [comment]
Jotai core is probably unopinionated about it. Some possible options are:

- https://github.com/jotaijs/jotai-tanstack-query @kalijonn 
- https://github.com/jotaijs/jotai-ssr @akkadaska
- or a new package

**Reply by akkadaska** [reply]
@phuctm97 

`jotai-ssr` provides a `HydrationBoundary` component inspired by Tanstack Query.

For more details, check out the README section:
https://github.com/jotaijs/jotai-ssr?tab=readme-ov-file#hydrating-atoms

I’m currently working on refactoring, and the updates will be released soon. The new README explains the component in detail and how to hydrate atoms using this library.
You can preview it here:
https://github.com/jotaijs/jotai-ssr/pull/11
https://github.com/jotaijs/jotai-ssr/blob/4f2325bc0110507336a10510c90275a52d5df00c/README.md#hydrationboundary

If you have any ideas, feel free to open an issue or create a pull request!
https://github.com/jotaijs/jotai-ssr
